Merge branch-Sort (Developer Guide) with Master (#100)
AY2223S1-CS2103-F09-2/tp: docs/DeveloperGuide.md

This section describes some noteworthy details on how certain features are implemented.

### Sort Persons

The sort persons feature allows the user to sort persons by their name, date of birth or gender. Sorting is performed as part of `listPersonsCommand#execute()`.

The sorting operation is exposed in the `Model` interface as `Model#sortPersons()` which calls `AddressBook#sortPersons()` which in turn calls `UniquePersonList#sort()` to sort the underlying `ObservableList<Person>`.

The following sequence diagram illustrates the relevant sorting method calls for the command `listPersons s/n` with a `SortField` object `s`:

![SortPersonsSequenceDiagram](images/SortPersonsSequenceDiagram.png)

The field to sort persons by is encapsulated within the `SortField` class. `SortField` stores the sort field enumeration (`SortFieldType`) as well as the `Comparator` object that defines the sorting logic.

The following class diagram shows how the `SortField` class integrates with the other components:

> ![SortFieldClassDiagram](images/SortFieldClassDiagram.png)

Reviewer: It may be better to present the SortField class and its class diagram before the sequence diagram, as I was initially confused as to where the SortField class came from.

> <div markdown="span" class="alert alert-info">:information_source: **Note:** Sorting is performed directly on the underlying `UniquePersonList` object, which means the sorted result is **permanent**. i.e. Not a temporary result for the `listPersons` command. For example, if `listPersons s/n` and `listPersons` are executed back-to-back, the result of the second `listPersons` command will still display the sorted results from the first `listPersons s/n` command because the sorted result is permanent.<br><br>

Reviewer: ```suggestion
<div markdown="span" class="alert alert-info">:information_source: **Note:** Sorting is performed directly on the underlying `UniquePersonList` object, which means the sorted result is **permanent**. For example, if `listPersons s/n` and `listPersons` are executed back-to-back, the result of the second `listPersons` command will display the sorted results from the first `listPersons s/n` command because the sorted result is permanent.<br><br>
```
I was thinking these changes might make the documentation less repetitive!

>    Ideally, sorting should be a "view" level operation that doesn't change the underlying persons list. However, the way the UI is designed makes it difficult to implement sorting as such. The issue is that the UI is **hardcoded** to only display persons from the `UniquePersonList` object. As such, any changes to the persons list must be made directly to the `UniquePersonList` object. i.e. A modified copy of the persons list cannot be passed to the UI during runtime since the UI is hardcoded to show only the `UniquePersonList` object.<br><br>

Reviewer: Hmmm... I would add this under Design Considerations as this issue was something that was raised up during our previous meetings, although I can understand that it would conflict with the listEvents DG because the point raised might be the same, I believe that since this is an individual DG, it would be fine to do so.

>    Therefore given the constraints of the UI, we have to implement sorting as a **permanent operation**.

Reviewer: ```suggestion
   Therefore, given the constraints of the UI, sorting is implemented as a **permanent operation**.
```
Some grammar improvements that I'd recommend
